File: algorithms/C/sorting/bubble-sort.c
//bubble sort
#include <stdio.h>
#include <stdlib.h>

void swap(int *value1, int *value2);

void bubble_sort(int array[], int array_size)
{
    for(int i = 0; i < array_size; i++)
    {
        for(int j = 0; j < array_size - 1; j++)
        {
            if(array[j] > array[j+1])
            {
                swap(&array[j], &array[j+1]);
            }
        }
    }
}

void swap(int *value1, int *value2)
{
    int temp = *value1;
    *value1 = *value2;
    *value2 = temp;
}

int main()
{
    int array_size;
    do{
        printf("Input the size of the array: ");
        scanf("%d", &array_size);
    }while(array_size <= 1);
    
    int *array = (int*)malloc(sizeof(int)*array_size);
    for(int i = 0; i < array_size; i++)
    {
        printf("Input an integer: ");
        scanf("%d", &array[i]);
    }
    
    bubble_sort(array, array_size);
    
    printf("Sorted array: [ ");
    for(int i = 0; i < array_size; i++)
    {
        printf("%d ", array[i]);
    }
    printf("]\n");
    return 0;
}

/*
 * Input the size of the array: 5 
 * Input 5 integers: [ 28, 17, 2, 5, 43 ]
 * Output: Sorted array: [ 2, 5, 17, 28, 43 ]
 *
 * Time complexity: O(n^2)
 */

File: algorithms/Go/sorting/bubble-sort.go
/*
   Bubble sort is a recursive algorithm based on swapping 2 values closest to each other: k and k+1
   We start by checking the first two values of the array, and we swap them if the first value is bigger than the second one.
   Then we recursively swap the second value with the third, if needed and so on until the array is sorted.
   Average Time Complexity: O(n^2))
*/

package sorting

import (
	"fmt"
)

func bubbleSort(sliceInt []int, n int) {
	for k := 0; k < n-1; k++ {
		if sliceInt[k] > sliceInt[k+1] {
			// We swaps the value if sliceInt[k] > sliceInt[k+1]
			sliceInt[k], sliceInt[k+1] = sliceInt[k+1], sliceInt[k]
			bubbleSort(sliceInt, n)
		}
	}
}

func RunBubbleSort() {
	sliceInt := []int{1, 2, -1, 0, 534, -100, 9, 53, 203}
	bubbleSort(sliceInt, len(sliceInt))
	fmt.Println(sliceInt)
}

